sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,358 @@
+    op match {
+      case EqualTo(l, r) => evaluateEqualTo(op, planStat, attrRef, literal, update)
+      case _ =>
+        attrRef.dataType match {
+          case StringType => evaluateBinaryForString(op, planStat, attrRef, literal, update)
+          case dataType: DataType if (dataType.isInstanceOf[NumericType]
+            || dataType.isInstanceOf[DateType]
+            || dataType.isInstanceOf[TimestampType]) =>
+            evaluateBinaryForNumeric(op, planStat, attrRef, literal, update)
+        }
+    }
+  }
+
+  /** This method evaluates the equality predicate for all data types. */
+  def evaluateEqualTo(
+      op: BinaryComparison,
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+
+    var percent: Double = 1.0
+    val aColStat = planStat.colStats(attrRef.name)
+    val ndv = aColStat.distinctCount
+    val datumString = attrRef.dataType match {
+      case d: DateType if literal.dataType.isInstanceOf[StringType] =>
+        val date = DateTimeUtils.stringToDate(literal.value.asInstanceOf[UTF8String])
+        if (date.isEmpty) {
+          logInfo("[CBO] Date literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+        date.get.toString
+      case t: TimestampType if literal.dataType.isInstanceOf[StringType] =>
+        val timestamp = DateTimeUtils.stringToTimestamp(literal.value.asInstanceOf[UTF8String])
+        if (timestamp.isEmpty) {
+          logInfo("[CBO] Timestamp literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+        timestamp.get.toString
+      case _ => literal.value.toString
+    }
+
+    val minString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+    val maxString: String = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.max match {
+          case Some(v) => v.toString
+          case None => ""
+        }
+      case _ => ""
+    }
+
+    /**
+     * decide if the value is in [min, max] of the column.
+     * We currently don't store min/max for binary/string type.
+     * Hence, we assume it is in boundary for binary/string type.
+     */
+    val inBoundary: Boolean = attrRef.dataType match {
+      case dataType: DataType if (dataType.isInstanceOf[NumericType]
+        || dataType.isInstanceOf[DateType]
+        || dataType.isInstanceOf[TimestampType]) =>
+        aColStat.min match {
+          case Some(v) =>
+            datumString.toDouble >= minString.toDouble && datumString.toDouble <= maxString.toDouble

```

Comment:
I don't think it's ok to convert everything to string and then to double, it will cause precision loss and inefficiency. It's ok in our cbo prototype but not in spark.